### ISSUE TITLE
HPA development is not active

### ISSUE THREAD

**Issue by sanposhiho** [kubernetes/kubernetes#128948]
/sig autoscaling
/cc @kubernetes/sig-autoscaling-misc 

## What

HPA's development has not been active recently.
It causes many PRs to struggle to get reviews, including some KEPs.
Essentially, this is the problem of lacking approvers in HPA.

## Context (AFAIK)

Currently, @mwielgus is the only approver, but already left the sig-autoscaling chair, and I'm not sure if they're still willing to help in reviewing.
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/podautoscaler/OWNERS#L7
I know the step down doesn't always mean a stop working on things, but, in this case, actually the last review from them was more than 1 year ago.
https://github.com/kubernetes/kubernetes/pulls?q=is%3Apr+reviewed-by%3Amwielgus+is%3Aclosed

There are some minor changes made in HPA though, all of them are approved by someone else (root approvers), not stamps from sig-autoscaling.
https://github.com/kubernetes/kubernetes/commits/master/pkg/controller/podautoscaler

Also, @gjtempleton is trying to take over the position, and he's (only) a reviewer (not yet approver) apart from @mwielgus now.
- https://github.com/kubernetes/kubernetes/pull/124607
- https://github.com/kubernetes/kubernetes/pull/124661

## Proposal

We're trapped in a vicious cycle; HPA development is not active because of the lack of reviewers/approvers, and no reviewer is newly born because HPA development is not active.

There's (probably, AFAIK) no other active person who is eligible for the reviewer/approver of HPA based on [the official criteria](https://github.com/kubernetes/community/blob/master/community-membership.md). 
But, we shouldn't keep the current situation, and I'd propose having some volunteers to join reviewers/approvers to break through (even if they're not officially eligible).

Regarding the approver, I cannot come up with any idea other than asking @gjtempleton to be an approver and start approving some PRs. (... I know they're also busy though)  
Also, when I was doing the container-based HPA enhancement, I remember @pbarker also helped reviewing a lot, might be a good idea to ask them to join the reviewer list. ([PRs](https://github.com/kubernetes/kubernetes/pulls?q=is%3Apr+author%3Apbetkier+is%3Aclosed), [reviews](https://github.com/kubernetes/kubernetes/pulls?q=is%3Apr+reviewed-by%3Apbetkier+is%3Aclosed+label%3Asig%2Fautoscaling)) I can also help in being a reviewer too. ([PRs](https://github.com/kubernetes/kubernetes/pulls?q=is%3Apr+author%3Asanposhiho+is%3Aclosed+label%3Asig%2Fautoscaling), [reviews](https://github.com/kubernetes/kubernetes/pulls?q=is%3Apr+reviewed-by%3Asanposhiho+is%3Aclosed+label%3Asig%2Fautoscaling+))

**Comment by adrianmoisey** [comment]
This problem also applies to the VPA.
Myself (and a small group of others, @omerap12 being one of them) are trying to contribute to the VPA, in order to climb the ranks to become approvers. But it is catch 22, since the current approvers are very busy and can't dedicate much time to review.
For the moment we try review amongst ourselves, getting PRs into a good state. Then we batch up a bunch of PRs and ask the approvers to approve. This is mostly working well, but will take some time before we get to the point where we have more approvers.

I'm happy to turn my attention to the HPA too, but for now all I can offer is code review, no approval.

**Comment by omerap12** [comment]
> This problem also applies to the VPA. Myself (and a small group of others, @omerap12 being one of them) are trying to contribute to the VPA, in order to climb the ranks to become approvers. But it is catch 22, since the current approvers are very busy and can't dedicate much time to review. For the moment we try review amongst ourselves, getting PRs into a good state. Then we batch up a bunch of PRs and ask the approvers to approve. This is mostly working well, but will take some time before we get to the point where we have more approvers.
> 
> I'm happy to turn my attention to the HPA too, but for now all I can offer is code review, no approval.

I agree with the above. I'm also willing to help, but I'm not very familiar with the HPA codebase yet (though I'd like to dive into it).

**Comment by DP19** [comment]
I can also volunteer some time here; I've become pretty familiar with the code when comparing it against a GKE issue we ran into [here](https://issuetracker.google.com/issues/309958646) and also working on this [bug](https://github.com/kubernetes/kubernetes/pull/127193) (that's still waiting to be approved like you've mentioned)

**Comment by wzshiming** [comment]
This problem also bothered me, and I tried to push the from/to 0 feature of HPV but didn't get much progress

**Comment by soltysh** [comment]
This was brought to my attention as one of the steering members [overlooking  sig-autoscaling](https://github.com/kubernetes/community/blob/master/liaisons.md#liaisons). I've started a conversation with current and past leads to ensure we can move forward and get us all in a situation that will ensure the SIG has the necessary power to approve code changes.

**Comment by sftim** [comment]
Is this possibly also relevant to SIG ContribEx?

**Comment by sftim** [comment]
>  Myself (and a small group of others, @omerap12 being one of them) are trying to contribute to the VPA, in order to climb the ranks to become approvers. But it is catch 22, since the current approvers are very busy and can't dedicate much time to review.

BTW, SIGs can choose to treat _opening_ of reasonable-looking PRs as grounds to accept a new reviewer.

Reviewers should try to make sure that the queue of ready-to-approve PRs is tidy and that only PRs genuinely ready for approval reach that stage.

**Comment by BenTheElder** [comment]
@soltysh we should check in again here.

I agree that people can sign up for reviewer immediately. Reviewer is a way to help be assigned initial reviews before approvers and gain knowledge. If anyone is interested that is a great starting point.

**Comment by raywainman** [comment]
I wonder if we can close this now?

With https://github.com/kubernetes/org/pull/5503 just waiting for code freeze to lift before submission we will now have more reviewers + owners in place here.

We've also made a few contributions to HPA in the last little while, for example the [tolerance feature](https://github.com/kubernetes/kubernetes/pull/130797) that went out in 1.33. I expect momentum to ramp up here.

cc @jackfrancis @gjtempleton @towca

**Comment by gjtempleton** [comment]
@raywainman I'd say lets raise the PR to add yourself and @adrianmoisey to the reviewers list under OWNERS as well.

**Comment by omerap12** [comment]
> @raywainman I'd say lets raise the PR to add yourself and @adrianmoisey to the reviewers list under OWNERS as well. 
> 
> Edit: @omerap12 as well, assuming he's still wanting to join in?

Count me in :)

**Comment by towca** [comment]
> With [kubernetes/org#5503](https://github.com/kubernetes/org/pull/5503) just waiting for code freeze to lift before submission we will now have more reviewers + owners in place here.

Are we sure that this will result in a meaningful increase to PR reviewing bandwidth for HPA approvers? I talked to @adrianmoisey about this during KubeCon, and IIUC (please correct me if I got it wrong Adrian) he had trouble finding approvers for HPA PRs fairly recently.

**Comment by adrianmoisey** [comment]
> > With [kubernetes/org#5503](https://github.com/kubernetes/org/pull/5503) just waiting for code freeze to lift before submission we will now have more reviewers + owners in place here.
> 
> Are we sure that this will result in a meaningful increase to PR reviewing bandwidth for HPA approvers? I talked to @adrianmoisey about this during KubeCon, and IIUC (please correct me if I got it wrong Adrian) he had trouble finding approvers for HPA PRs fairly recently.

That is correct, however, this issue (and the suggestions in it) is the path forward to getting approvers (over time)

**Comment by towca** [comment]
> > > With [kubernetes/org#5503](https://github.com/kubernetes/org/pull/5503) just waiting for code freeze to lift before submission we will now have more reviewers + owners in place here.
> > 
> > 
> > Are we sure that this will result in a meaningful increase to PR reviewing bandwidth for HPA approvers? I talked to [@adrianmoisey](https://github.com/adrianmoisey) about this during KubeCon, and IIUC (please correct me if I got it wrong Adrian) he had trouble finding approvers for HPA PRs fairly recently.
> 
> That is correct, however, this issue (and the suggestions in it) is the path forward to getting approvers (over time)

Understood, in this case I'd keep the issue open until the situation improves (and e.g. we have 2 active HPA approvers). @raywainman WDYT?

**Comment by soltysh** [comment]
> [@soltysh](https://github.com/soltysh) we should check in again here.
> 
> I agree that people can sign up for reviewer immediately. Reviewer is a way to help be assigned initial reviews before approvers and gain knowledge. If anyone is interested that is a great starting point.

During KubeCon EU I've talked with several people who are working on progressing through the contributor ladder, and I'm actively in contact with them. It's moving forward, although it will take a few more weeks, maybe even months 🤞 

> With [kubernetes/org#5503](https://github.com/kubernetes/org/pull/5503) just waiting for code freeze to lift before submission we will now have more reviewers + owners in place here.

This changes leads, which is parallel effort. I don't have any objections to closing this issue. Like I wrote above work is happening and I'm personally involved in helping sig-autoscaling earn the approver rights in hpa. 

> Are we sure that this will result in a meaningful increase to PR reviewing bandwidth for HPA approvers? I talked to [@adrianmoisey](https://github.com/adrianmoisey) about this during KubeCon, and IIUC (please correct me if I got it wrong Adrian) he had trouble finding approvers for HPA PRs fairly recently.

As mentioned before the newly appointed leads are in-sync with me, so every PR they approve is then sent over to me for actual approval label. This way they are earning the necessary confidence to become approvers. I might have talked with Adrian during KubeCon, b/c I clearly remember talking with someone from South Africa 😅 but that might have been someone else.

**Comment by raywainman** [comment]
> Understood, in this case I'd keep the issue open until the situation improves (and e.g. we have 2 active HPA approvers). [@raywainman](https://github.com/raywainman) WDYT?

I'm good with that!

I've opened https://github.com/kubernetes/kubernetes/pull/131292 here to refresh the reviewers/approvers list. Please chime-in on that PR if I've missed anything.

Thanks all!

**Comment by adrianmoisey** [comment]
> As mentioned before the newly appointed leads are in-sync with me, so every PR they approve is then sent over to me for actual approval label. This way they are earning the necessary confidence to become approvers. I might have talked with Adrian during KubeCon, b/c I clearly remember talking with someone from South Africa 😅 but that might have been someone else.

Yup, that was me :)

**Comment by BenTheElder** [comment]
https://github.com/kubernetes/kubernetes/pull/131650 has gone in since the last updates here

**Comment by soltysh** [comment]
We're on a good path towards raising additional reviewers and approvers, I'm working closely with several folks from sig-autoscaling. I'm very happy with their work and progress.
